File: library/src/main/java/com/trinity/recording/processor/AudioRecordProcessor.java
package com.trinity.recording.processor;

public class AudioRecordProcessor implements RecordProcessor {
	private long handle;

	@Override
	public void initAudioBufferSize(int audioSampleRate, int audioBufferSize) {
		handle = init(audioSampleRate, audioBufferSize);
	}

	@Override
	public void pushAudioBufferToQueue(short[] audioSamples, int audioSampleSize) {
		pushAudioBufferToQueue(handle, audioSamples, audioSampleSize);
	}

	@Override
	public void flushAudioBufferToQueue() {
		flushAudioBufferToQueue(handle);
	}

	private native long init(int audioSampleRate, int audioBufferSize);

	private native void flushAudioBufferToQueue(long handle);
	private native void destroy(long handle);

	private native int pushAudioBufferToQueue(long handle, short[] audioSamples,
			int audioSampleSize);

	@Override
	public void destroy() {
		destroy(handle);
	}

}

File: library/src/main/java/com/trinity/record/processor/AudioRecordProcessor.kt
package com.trinity.record.processor

class AudioRecordProcessor : RecordProcessor {
    private var handle: Long = 0

    override fun initAudioBufferSize(audioSampleRate: Int, audioBufferSize: Int, speed: Float) {
        handle = init(audioSampleRate, audioBufferSize, speed)
    }

    override fun pushAudioBufferToQueue(audioSamples: ShortArray, audioSampleSize: Int) {
        pushAudioBufferToQueue(handle, audioSamples, audioSampleSize)
    }

    override fun flushAudioBufferToQueue() {
        flushAudioBufferToQueue(handle)
    }

    override fun destroy() {
        if (handle == 0L) {
            return
        }
        destroy(handle)
        handle = 0
    }

    private external fun init(audioSampleRate: Int, audioBufferSize: Int, speed: Float): Long
    private external fun flushAudioBufferToQueue(handle: Long)
    private external fun destroy(handle: Long)
    private external fun pushAudioBufferToQueue(
        handle: Long, audioSamples: ShortArray,
        audioSampleSize: Int
    ): Int
}
